rmf_task/include/rmf_task/Estimate.hpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef RMF_TASK__ESTIMATE_HPP
+#define RMF_TASK__ESTIMATE_HPP
+
+#include <optional>
+#include <utility>
+
+#include <rmf_task/agv/State.hpp>
+#include <rmf_traffic/Time.hpp>
+#include <rmf_utils/impl_ptr.hpp>
+
+namespace rmf_task {
+
+/// Estimates for requests
+class Estimate
```

Comment:
I overlooked this concern in my first review, but is there any need for users to have access to this `Estimate` class? If it's only used internally by the task allocation planner, then there's no need for it to be in the public API. In that case, this header should be moved to `rmf_task/src/rmf_task/`. We should always try to minimize the surface area of the public API.

Reply:
An `Estimate` object is returned by the `estimate_finish()` [method](https://github.com/osrf/rmf_core/blob/89a68c0f0bbc8a26994c302c37f9d275713d141a/rmf_task/include/rmf_task/Request.hpp#L44-L47) in `Request` and hence is exposed in the public API.

Reply:
Ah, I see. How about `EstimateCache`? I'm not seeing where that's used by the public API.

Reply:
`EstimateCache` is an argument in the `estimate_finish()` function. It is also returned by [this](https://github.com/osrf/rmf_core/blob/89a68c0f0bbc8a26994c302c37f9d275713d141a/rmf_task/include/rmf_task/agv/TaskPlanner.hpp#L192-L193) function in the `TaskPlanner` API which gets called in `TaskManager.cpp` [here](https://github.com/osrf/rmf_core/blob/89a68c0f0bbc8a26994c302c37f9d275713d141a/rmf_fleet_adapter/src/rmf_fleet_adapter/TaskManager.cpp#L391)

Reply:
I'm somewhat unclear on the exact lifecycle of the `EstimateCache`, i.e. who is responsible for creating it and how it gets passed around to different objects. Generally I would expect any type of "cache" to be an implementation detail which is hidden from the public API. Ideally users shouldn't have to be aware that we're doing any kind of caching inside the library.

That being said, you've made it clear to me that it is currently a significant item in the public API, so we can move forward with it as-is and revisit this concern before the next release :+1: